samza-azure/src/main/java/org/apache/samza/system/eventhub/consumer/EventHubSystemConsumer.java

```diff
@@ -0,0 +1,342 @@
+  private final Map<String, Serde<byte[]>> serdes;
+  private boolean isStarted = false;
+  private final EventHubConfig config;
+  private final String systemName;
+
+
+  public EventHubSystemConsumer(EventHubConfig config, String systemName,
+                                SamzaEventHubClientFactory samzaEventHubClientFactory,
+                                Map<String, Serde<byte[]>> serdes,  MetricsRegistry registry) {
+    super(registry, System::currentTimeMillis);
+
+    this.config = config;
+    this.systemName = systemName;
+    this.serdes = serdes;
+    List<String> streamNames = config.getStreams(systemName);
+    // Create and initiate connections to Event Hubs
+    for (String streamName : streamNames) {
+      SamzaEventHubClient ehClientWrapper = samzaEventHubClientFactory
+              .getSamzaEventHubClient(systemName, streamName, config);
+      streamEventHubClients.put(streamName, ehClientWrapper);
+      ehClientWrapper.init();
+    }
+
+    // Initiate metrics
+    eventReadRates = streamNames.stream()
+            .collect(Collectors.toMap(Function.identity(), x -> registry.newCounter(x, EVENT_READ_RATE)));
+    eventByteReadRates = streamNames.stream()
+            .collect(Collectors.toMap(Function.identity(), x -> registry.newCounter(x, EVENT_BYTE_READ_RATE)));
+    readLatencies = streamNames.stream()
+            .collect(Collectors.toMap(Function.identity(), x -> new SamzaHistogram(registry, x, READ_LATENCY)));
+    readErrors = streamNames.stream()
+            .collect(Collectors.toMap(Function.identity(), x -> registry.newCounter(x, READ_ERRORS)));
+
+    // Locking to ensure that these aggregated metrics will be created only once across multiple system consumers.
+    synchronized (AGGREGATE) {
+      if (aggEventReadRate == null) {
+        aggEventReadRate = registry.newCounter(AGGREGATE, EVENT_READ_RATE);
+        aggEventByteReadRate = registry.newCounter(AGGREGATE, EVENT_BYTE_READ_RATE);
+        aggReadLatency = new SamzaHistogram(registry, AGGREGATE, READ_LATENCY);
+        aggReadErrors = registry.newCounter(AGGREGATE, READ_ERRORS);
+      }
+    }
+  }
+
+  @Override
+  public void register(SystemStreamPartition systemStreamPartition, String offset) {
+    super.register(systemStreamPartition, offset);
+
+    if (isStarted) {
+      throw new SamzaException("Trying to add partition when the connection has already started.");
+    }
+
+    if (StringUtils.isBlank(offset)) {
+      switch (config.getStartPosition(systemName, systemStreamPartition.getStream())) {
+        case EARLIEST:
+          offset = START_OF_STREAM;
+          break;
+        case LATEST:
+          offset = END_OF_STREAM;
+          break;
+        default:
+          throw new SamzaException("Unknown starting position config " +
+                  config.getStartPosition(systemName, systemStreamPartition.getStream()));
+      }
+    }
+    streamPartitionOffsets.put(systemStreamPartition, offset);
```

Comment:
Please correct me if i'm wrong here.

I think this has to use offsetComparator and pick the lowest offset. 

If we've two tasks with checkpointed offsets T1(SP1: START_OF_STREAM), T2(SP1: END_OF_STREAM), register should choose the lowest one and store in streamPartitionOffsets for a partition(This register call will be called once for each TaskInstance). If we just overwrite, we could pick End (depending upon register call order).

Reply:
Good catch, I was relying on the `register` being only called once per SSP